pxemgr/pxemanager.go

```diff
@@ -118,19 +121,49 @@ func PXEManager(c PXEManagerConfiguration, cluster *hostmgr.Cluster) (*pxeManage
 		mu: new(sync.Mutex),
 	}
 
+	// check for deprecated EtcdDiscoveryUrl
+	if mgr.cluster.Config.EtcdDiscoveryURL != "" && mgr.cluster.Config.DefaultEtcdClusterToken == "" {
+		// transform discovery url to token
+		parts := strings.Split(mgr.cluster.Config.EtcdDiscoveryURL, "/")
+		token := parts[len(parts)-1]
+		baseUrl := strings.Join(parts[:len(parts)-1], "/")
+
+		if mgr.useInternalEtcdDiscovery {
+			// convert token to internal etcd discovery
+			err := mgr.cluster.StoreEtcdDiscoveryToken(mgr.etcdEndpoint, token, mgr.defaultEtcdQuorumSize)
+			if err != nil {
+				glog.Fatalf("Can't store discovery token in etcd.", baseUrl, mgr.etcdDiscoveryUrl)
+			}
+
+			glog.Warningf("Transferred etcd token to internal discovery. Note that your machines still have the old discovery url in their cloud-config and that you need to transfer the current member data yourself.")
+		} else if mgr.etcdDiscoveryUrl != baseUrl {
+			glog.Fatalf("Deprecated EtcdDiscoveryURL ('%s') in your cluster.json differs from the --etcd-discovery parameter ('%s').", baseUrl, mgr.etcdDiscoveryUrl)
+		}
+		mgr.cluster.Config.EtcdDiscoveryURL = ""
+		mgr.cluster.Config.DefaultEtcdClusterToken = token
+		mgr.cluster.Commit(fmt.Sprintf("Convert deprecated etcd discovery url to default etcd token '%s'", token))
+	}
+
 	if mgr.cluster.Config.DefaultEtcdClusterToken == "" {
 		var (
 			token string
 			err   error
 		)
 		if mgr.useInternalEtcdDiscovery {
 			mgr.etcdDiscoveryUrl = mgr.thisHost() + "/etcd"
-			token, err = mgr.cluster.GenerateEtcdDiscoveryToken(mgr.etcdEndpoint, mgr.defaultEtcdQuorumSize)
+			token, err = mgr.cluster.GenerateEtcdDiscoveryToken()
+			if err != nil {
+				glog.Fatalf("Failed to generate etcd cluster token: %s", err)
+			}
+			err := mgr.cluster.StoreEtcdDiscoveryToken(mgr.etcdEndpoint, token, mgr.defaultEtcdQuorumSize)
```

Comment:
The size of the cluster is configured in etcd per token. This can be done via `curl -X PUT localhost:4080/etcd/new -d size=1`. If size is omitted the default value is used. You can query the size in etcd itself `etcdctl get  /_etcd/registry/58a60912bc9ea999fc6f4c700173d610/_config/size`. All cluster information itself is stored in etcd. I decided against duplicate state in Mayu for now. So etcdctl or curl is your friend here.